### DISCUSSION TITLE
Clarification on the variance of `self` on Protocols

### DISCUSSION THREAD

**Discussion by MathiasSven** [microsoft/pyright#3373] [General]
I am not sure pyright is the right place to open this as a bug report as I am not sure what the consensus is on the expected behavior, regardless to me it seems that when annotating a `Protocol` this construction should be allowed:
```py
class A(Protocol[_T_co]):
     def foo(self: _T_co): ...
```
Currently mypy accepts this, however pyright complains that since the type variable is used as a parameter it ought to be contravariant, conversely if you set the type variable to contravariant, then mypy complains about the opposite. Due to the nature of `self` should it not be treated as  covariant?

As to how this could be useful, I think broadly not so much, but it does is give you access to the type when annotating functions with such a protocol. In my specific use case:
```py
class A(Protocol[_T_co, _T_contra, _R_co]):
     def foo(self: _T_co, __other: _T_contra) -> _R_co: ...

def bar(x: A[_T, _T1, _T2], y: A[_T1, _T, _T3]) -> _T2 | _T3:
    a = x.foo(y) # Safe, will return _T2
    b = y.foo(x) # Safe, will return _T3
    
    if some_condition:
        return a
    else:
        return b
```

**Comment by erictraut** [comment]
Converting to discussion.

**Reply by erictraut** [reply]
In general, TypeVars that are used in parameters (and only in parameters) must be contravariant. I don't see why the type of `self` would be any different. You could ask in the [discussion section of the python/typing project](https://github.com/python/typing/discussions).

I would recommend against using the approach in the code above. I guess I don't fully understand your use case because I don't see type `_T_co` being used in any useful way within the protocol definition. Perhaps you could flesh out the sample to better explain what problem you're trying to solve?

Perhaps the new `Self` type recently introduced in [PEP 673](https://peps.python.org/pep-0673/) would help you out here?

**Reply by MathiasSven** [reply]
My goal here to see what can be done to the `operator` module in typeshed in varying levels of precision, where most of the functions are typed as `op(x: Any, y: Any) -> Any`

After looking a little further into it, even [using mypy](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=79e105971bb14f4352c2c91275bcbc43) where I am allowed to type the self parameter as covariant, that does not really solve my issue, I think I just dont fully understand how typevars resolve in this case:
```py
class SupportsAdd(Protocol[_T_co, _T_contra, _R_co]):
    def __add__(self: _T_co, __other: _T_contra) -> _R_co: ...

class SupportsLRAdd(Protocol[_T_co, _T_contra, _T2_contra, _R_co, _R2_co]):
    def __add__(self: _T_co, __other: _T_contra) -> _R_co: ...
    def __radd__(self: _T_co, __other: _T2_contra) -> _R2_co: ...


def add(x: SupportsAdd[_T, _T1, _T2], y: SupportsLRAdd[_T, _T1, _T, _T2, _T3]) -> _T2 | _T3: ...


class A:
    def __add__(self, __other: Any) -> A: ...

class SubA(A):
    def __radd__(self, __other: SubA) -> bool: ...


# Correctly erros (This is my estimate as to what y resolves to when calling `add(A(), SubA())`)
t: SupportsLRAdd[A, Any, A, A, bool] = SubA() 

reveal_type(add(A(), SubA()))  # Mypy still allows this
```
The `add` function here would just be one of the overloads of a fully typed `operator.add` function and I was hoping `add(A(), SubA())` would not be accepted by this overload since `SubA` does not accept `A` as an `__other` parameter on `__radd__`.

The rational behind using a `self` as a covariant typevar is that if `y` resolves to `SupportsLRAdd[A, Any, Any, Any, Any]` then `SubA` should be accepted by that since it would be a `SupportsLRAdd[SubA, Any, Any, Any, Any]`. The need to distinguish if the two operands are of the same class or not is an unfortune consequence of the quirk I raised an issues about recently, as you pointed out there is no way to correctly distinguish between a type and its subtype in this instances, however I was hopping to be able to return an Union of the results `iff` the types match, and otherwise give the more precise result that follows from the sensible order `__add__` first, if not, `__radd__`.

Independently of this, I believe that the `self` is different from other parameters in that it is inherently covariant, for example:
```py
class Foo:
    def bar(self, arg: int) -> int:
        return hash(arg)

class SubFoo(Foo):
    salt: int

    def __init__(self, salt: int) -> None:
        self.salt = salt

    def bar(self, arg: int) -> int:
        return hash(arg) ^ obj.salt
```
Here `SubFoo` follows LSP, however calling `SubFoo.bar(A(), 42)` would throw an error as `SubFoo.bar` requires that `self` be of type `SubFoo`. `SubFoo` is only a lawful subclass of `Foo` because wherever you can use an *instance* of `Foo`, an *instance* of `SubFoo` would also work, that is after `self` has been hidden away by method bounding.

I am not claiming that `SubFoo.bar` should be a subtype of `Foo.bar`, just that given the context of how method bounding works, it makes sense to allow `self` to be typed as covaraint if that has any use at all (which granted I have not been able to show). If pyright just pretended `self` didn't exist when comparing instances this would be less of a problem, but currently it seems that it does still take that into account:
```py
class A:
    def foo(self): ...

class SubA(A): ...

class ProtoA(Protocol[_T_contra]):
    def foo(self: _T_contra): ...

a: ProtoA[A] = SubA()  # Error in pyright 
```

**Reply by MathiasSven** [reply]
> After looking a little further into it, even using mypy where I am allowed to type the self parameter as covariant, that does not really solve my issue, I think I just don't fully understand how typevars resolve in this case:

Upon further inspection, the reason that mypy did not solve my issue appears to be a bug with mypy:
```py
class ProtoFoo(Protocol[_T_co]):
    def foo(self: _T_co) -> None: ...

class Foo:
    def foo(self) -> None: ...

class SubFoo(Foo): ...
    
a: ProtoFoo[SubFoo] = Foo()  # Fine, but shouldn't
```
Since mypy complains if `self` is typed anything other then covariant, I believe this is intended to be supported, so I might raise it as an issue on mypy. However, this kind of construction can't be used without consensus on the correct behavior among type checkers so I would still like to know your thoughts on this before raising the issue.

**Reply by erictraut** [reply]
Independent of the question about whether the type of the `self` parameter should be enforced as covariant or otherwise, I don't think this pattern is valid. The problem is that you are creating a protocol where the implied type of `self` is the protocol class itself, but then you're making the protocol class generic and parameterizing it with a different type that is then applied to the `self` parameter. The correct type of `self` in this case is `ProtoFoo`. Annotating it as `_T_co` is somewhat nonsensical.

**Reply by erictraut** [reply]
Within non-protocol classes, pyright and mypy both enforce that a type annotation on `self` is a supertype of the containing class.

For example:
```python
class Parent:
    ...

class Child(Parent):
    def func1(self: Parent) -> None: # Allowed
        ...
    def func2(self: int) -> None:  # Error
        ...
    def func3(self: "Grandchild") -> None:  # Error
        ...

class Grandchild(Child):
    ...
```

This is consistent with the idea that the type of `self` is contravariant rather than covariant.

Neither pyright nor mypy enforce this "`self` type must be a supertype of the containing class" rule for protocol classes. That's because the notion of a "subtype" is more complex with structural subtyping, and recursion can easily result when determining whether one protocol is a subtype of another. But perhaps we should attempt to enforce this rule for protocol classes. If it were enforced, then `_T_co` would clearly be an illegal annotation for `self` because it is not a supertype of the containing class.

**Reply by MathiasSven** [reply]
> The problem is that you are creating a protocol where the implied type of `self` is the protocol class itself, but then you're making the protocol class generic and parameterizing it with a different type that is then applied to the `self` parameter. The correct type of `self` in this case is `ProtoFoo`. Annotating it as `_T_co` is somewhat nonsensical.

I do not see how this is nonsensical, a protocol is just a structural contract, if I have:
```py
class Foo: ...

class ProtoFoo(Protocol):
    def foo(self: Foo) -> Any: ...
``` 
`ProtoFoo` is a valid structural contract, where any type is a subtype of it iff it implements a foo method which takes `Foo` as a `self` parameter, the fact that `Foo` is actually not compatible with `ProtoFoo` should not be a concern of the protocol construction, but rather of the protocol assignment. A protocol never gets instantiated, it's methods are never bounded, there is no obligation that `self` is of type `ProtoFoo`, the only *implicit* obligation is that `self` be a subtype of `ProtoFoo`. The construction I am suggesting does not change the semantics of the behavior of a protocol nor does it seem make anything unsound (at least as I see it right now), it only adds the ability to further specify its subtype properties, eg.: "that it be a subtype of `Foo` as well".

> This is consistent with the idea that the type of self is contravariant rather than covariant.

I do not see how that example relates to covariance or contravariance, would that not just be a consequence of the fact that when you create an element of the type `Child` it will bind those methods with `Child` as the type for the `self` parameter, and `Child` is not compatible with `Grandchild` nor `int`? On the other hand, this shows the opposite, that "the first parameter on methods" is covariant:

```py
class A: ...
class SubA(A): ...
 
def foo(self: SubA, x: int) -> Any: ...

a: Callable[[A, int], Any] = foo  # Error as Callable[[SubA, int], Any] is not a subtype of Callable[[A, int], Any]

class Parent:
    def func1(self: 'Parent', x: int) -> Any: ...
    def func2(self: 'Parent', x: int) -> Any: ...

class Child(Parent):
    def func1(self: 'Child', x: int) -> Any: ...  # This is Allowed, it does not make anything unsound, 
                                                  # however were it not a method, this assignemnt would not be allowed.

    def func2(self: 'Child', x: bool) -> Any: ...  # This is an error due to contravaraince, but it only applies 
                                                   # to the parameters that are not bounded
```

**Reply by MathiasSven** [reply]
Regardless of this, I think whatever a typed self parameter on a Protocol would achieve, would also be achievable by a generic intersection of a protocol and a nominal class, when or if that ever becomes a thing.
